refactor: Select and Include type generation

SevInf opened the pull request:
Split type generation from `Model.ts` and use ts-builders for
generation. Adds several helpers, necessary for this refactoring:
`typeDeclaration`, `export` and `keyType`.

SevInf committed:
refactor: Select and Include type generation

Split type generation from `Model.ts` and use ts-builders for
generation. Adds several helpers, necessary for this refactoring:
`typeDeclaration`, `export` and `keyType`.

SevInf commented on packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/__snapshots__/test.ts.snap:
```diff
@@ -1736,9 +1736,7 @@ export namespace Prisma {
-  export type EmbedInclude = {
-
-  } 
+  export type EmbedInclude = {}
```
Whitespace changes only

aqrln commented on packages/client/src/generation/ts-builders/KeyType.ts:
```diff
@@ -0,0 +1,34 @@
+export class KeyType implements BasicBuilder {
+  constructor(public baseType: AnyTypeBuilder, public key: string) {}
+  write(writer: Writer): void {
+    const needsBraces =
```
Should this be something like `hasBraces` or `needsParens` instead?

Also, would it make sense to define this as a method in `BasicBuilder` interface and let each type tell if it needs to be wrapped when indexing, or is not worth it and not going to be useful elsewhere?

SevInf replied:
Thanks, did that in the next commit.

SevInf committed:
Address review comments

Introduce base class for all type declarations. Move method for wrapping
into parenthesis there, as well as `subKey` helper.

SevInf commented on packages/client/src/generation/ts-builders/TypeDeclaration.test.ts:
```diff
@@ -0,0 +1,26 @@
+})
+
+test('with generic parameters', () => {
+  const decl = typeDeclaration('B', A).addGenericParameter(genericParameter('T'))
```
Added.

SevInf committed:
Resolve circular dependency

aqrln reviewed (approved):
💯
